yellow-paper/docs/gas-and-fees/gas-metering-and-fees.md

```diff
@@ -0,0 +1,683 @@
+
+Instead, using a fixed cost plus scaling on actual resources consumed dramatically simplifies calculations, is a reasonable approximation, and affords abstraction if/when we transition to a different DA solution.
+:::
+
+
+# User transaction variables
+
+The user must be able to specify:
+
+- `da_gas_limit`: the **maximum** amount of DA gas the user is **willing** to pay for, **inclusive** of `teardown_da_gas`
+- `teardown_da_gas`: the amount of DA gas the user **will pay** for the teardown function
+- `max_fee_per_da_gas`: the maximum price the user is willing to pay per unit of DA gas in `$AZT`
+- `l2_gas_limit`: the **maximum** amount of l2 gas the user is **willing** to pay for, **inclusive** of `teardown_l2_gas`
+- `teardown_l2_gas`: the amount of l2 gas the user **will pay** for the teardown function
+- `max_fee_per_l2_gas`: the maximum price the user is willing to pay per unit of l2 gas in `$AZT`
+- `l1_gas_limit`: the **maximum** amount of l1 gas the user is **willing** to pay for, **inclusive** of `teardown_l1_gas`
+- `teardown_l1_gas`: the amount of l1 gas the user **will pay** for the teardown function
+- `max_fee_per_l1_gas`: the maximum price the user is willing to pay per unit of l1 gas in `$AZT`
+- `inclusion_fee`: the flat fee the user is willing to pay for transaction inclusion in `$AZT`
+
+
+
+# Sequencer checks
+
+Even though fees are not enforced by the protocol, the sequencer can check ahead of time if it is possible that the transaction **could** pay the fee.
+
+This is done by inspecting the public call enqueued for teardown, checking who the message sender is, and inspecting the balance of the `$AZT` for that sender.
+
+Note: it does not know by inspecting the enqueued call itself if the `pay_fee` function will be called, unless it has explicitly inspected/whitelisted the function ahead of time.
+
+However, it will know that if `pay_fee` is called, then message sender is the one that will pay the fee, and thus must have a balance greater than:
+```
+msg_sender.balance >= max_fee_per_da_gas * (da_gas_limit + teardown_da_gas)
+                        + max_fee_per_l2_gas * (l2_gas_limit + teardown_l2_gas)
+                        + max_fee_per_l1_gas * (l1_gas_limit + teardown_l1_gas)
+                        + inclusion_fee
+```
+
+It will likely also want to check that `max_fee_per_*_gas` are at least as high as the current `fee_per_*_gas`.
+
+# Tracking in kernels
+
+### GasUsed
+
+We have a `GasUsed` that contains:
+
+- `da_gas_used`
+- `l2_gas_used`
+- `l1_gas_used`
+
+### GasLimits
+
+We have a `GasLimits` that contains:
+
+- `da_gas_limit`
+- `max_fee_per_da_gas`
+- `teardown_da_gas`
+- `l2_gas_limit`
+- `max_fee_per_l2_gas`
+- `teardown_l2_gas`
+- `l1_gas_limit`
+- `max_fee_per_l1_gas`
```

Comment:
I'd suggest separating out these gas prices from the gas limits (or renaming the `GasLimits` struct)